features/discovery/src/main/java/org/opennms/netmgt/discovery/DiscoveryTaskExecutorImpl.java

```diff
@@ -147,42 +158,138 @@ public void run() {
+
+        CompletableFuture<DiscoveryResult> future = new CompletableFuture<>();
+        List<Detector> detectors = configFactory.getListOfDetectors(inetAddress, location);
+        if (detectors.size() > 0) {
+            // Run all the detectors.
+            try {
+                List<CompletableFuture<Boolean>> futures = detectors.stream().map(detector ->
+                        detect(detector, inetAddress, location)).collect(Collectors.toList());
+                // Combine all futures.
+                CompletableFuture<Void> allFutures = CompletableFuture.allOf(futures.toArray(new CompletableFuture[futures.size()]));
+                CompletableFuture<List<Boolean>> futureList = allFutures.thenApply(result -> {
+                    return futures.stream().map(CompletableFuture::join).collect(Collectors.toList());
+                });
+                future = futureList.thenApply((results) -> {
+                    boolean allResult = results.stream().allMatch(result -> result);
+                    return new DiscoveryResult(allResult, inetAddress, pingDuration);
+                });
+                return future;
+            } catch (Exception e) {
+                LOG.error("Exception while performing detection in discovery.", e);
+                future.complete(new DiscoveryResult(false, inetAddress, pingDuration));
+            }
+        }  else {
+            //When there are no detectors, The discovery is just ping itself, make DiscoveryResult successful.
+            future.complete(new DiscoveryResult(true, inetAddress, pingDuration));
+        }
+        return future;
+    }
+
+    /**
+     * This calls detection asynchronously for a given detector and returns a completable future.
+     */
+    private CompletableFuture<Boolean> detect(Detector detector, InetAddress inetAddress, String location) {
+
+        CompletableFuture<Boolean> future = new CompletableFuture<>();
+        try {
+            Map<String, String> attributes = detector.getParameters().stream()
+                    .collect(Collectors.toMap(Parameter::getKey, Parameter::getValue));
+            CompletableFuture<Boolean> result = getLocationAwareDetectorClient().detect()
+                    .withAddress(inetAddress)
+                    .withClassName(detector.getClassName())
+                    .withLocation(location)
+                    .withAttributes(attributes)
+                    .execute();
+            result.whenComplete((response, ex) -> {
+                if (ex != null) {
+                    // Don't throw exception as Future join won't like.
+                    LOG.error("Exception while detecting {} service on IP Address {} at location {} ",
+                            detector.getName(), inetAddress.getHostAddress(), location, ex);
+                    future.complete(false);
+                } else {
+                    LOG.info("Detected service {} on IP Address {} at location {}",
```

Comment:
`Detected` is misleading here, since the result may be `false`.

Reply:
This log message appears in `provisiond.log`, we should probably re-route it to `discovery.log`.